src/MongoDB.Driver/MongoDatabaseImpl.cs

```diff
@@ -191,6 +193,34 @@ public override IMongoCollection<TDocument> GetCollection<TDocument>(string name
             return new MongoCollectionImpl<TDocument>(this, new CollectionNamespace(_databaseNamespace, name), settings, _cluster, _operationExecutor);
         }
 
+        public override IAsyncCursor<string> ListCollectionNames(ListCollectionNamesOptions options = null, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            return UsingImplicitSession(session => ListCollectionNames(session, options, cancellationToken), cancellationToken);
+        }
+
+        public override IAsyncCursor<string> ListCollectionNames(IClientSessionHandle session, ListCollectionNamesOptions options = null, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            Ensure.IsNotNull(session, nameof(session));
+            var operation = CreateListCollectionNamesOperation(options);
+            var effectiveReadPreference = ReadPreferenceResolver.GetEffectiveReadPreference(session, null, ReadPreference.Primary);
+            var cursor = ExecuteReadOperation(session, operation, effectiveReadPreference, cancellationToken);
+            return new BatchTransformingAsyncCursor<BsonDocument, string>(cursor, ExtractCollectionNames);
```

Comment:
BatchTransformingAsyncCursor uses the ExtractCollectionNames method to suck out the "name" fields from the response.